### ISSUE TITLE
Gr00t running on SO-100 arm, but it fails to do anything.

### ISSUE THREAD

**Issue by johnny555** [NVIDIA/Isaac-GR00T#82]
### 🐛 Describe the bug

Hi There, 

Thanks for the cool VLA model. I'm trying to get it to work on my SO-100 arm, and following the guide, but it is not working. 

I suspect I have missed some important key step, but I don't know what it is.

The robot arm initialises but then just shudders (see video)

https://github.com/user-attachments/assets/37b0f5a7-25e9-4051-ae61-894683d152d3

I followed the guide here: https://github.com/NVIDIA/Isaac-GR00T/blob/main/getting_started/3_new_embodiment_finetuning.ipynb to perform fine tuning, on this dataset: https://huggingface.co/datasets/badwolf256/so100_twin_cam_duck_v3 although only using the realsense scene cam as I didn't have VRAM for both video streams (using a 3090). 

I used this `modality.json` file: 
```
{
    "state": {
        "single_arm": {
            "start": 0,
            "end": 5
        },
        "gripper": {
            "start": 5,
            "end": 6
        }
    },
    "action": {
        "single_arm": {
            "start": 0,
            "end": 5
        },
        "gripper": {
            "start": 5,
            "end": 6
        }
    },
    "video": {
        "webcam": {
            "original_key": "observation.images.realsense"
        }
    },
    "annotation": {
        "human.task_description": {
            "original_key": "task_index"
        }
    }
}
```

It seemed to have converged after two runs based on my tensorboard output: 

![Image](https://github.com/user-attachments/assets/73f903e1-dead-400c-8bb1-df4db9e5752f)

(Note I used lora with rank 64)

I then setup an inference server using the command: 
`python scripts/inference_service.py --server     --model_path ~/so100-checkpoints/     --embodiment_tag new_embodiment     --data_config so100     --denoising_steps 4`

I then used this script to control the robot (using lerobot environment)  :

https://github.com/NVIDIA/Isaac-GR00T/blob/main/getting_started/examples/eval_gr00t_so100.py with some modifications (namely changing the text prompt, adjusting the RGB colours because they seemed wrong). 

I also needed the script https://github.com/NVIDIA/Isaac-GR00T/blob/main/gr00t/eval/service.py in the same directory that eval_groot_so100.py was run from.

There aren't any errors being output. The system appears to be running fine, but the model is not attempting the task. 

What have a missed ? Maybe I need more data? Do I need to train even longer? I'm training and inferencing on a 3090. 

Thanks for your help! 

### Versions

My gr00t environment:
ython 3.10.16
absl-py==2.2.1
accelerate==1.2.1
aiosignal==1.3.2
albucore==0.0.17
albumentations==1.4.18
annotated-types==0.7.0
antlr4-python3-runtime==4.9.3
anyio==4.9.0
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
arrow==1.3.0
asttokens==3.0.0
astunparse==1.6.3
async-lru==2.0.5
attrs==25.3.0
av==12.3.0
babel==2.17.0
beautifulsoup4==4.13.3
bleach==6.2.0
blessings==1.7
cachetools==5.5.2
certifi==2025.1.31
cffi==1.17.1
charset-normalizer==3.4.1
click==8.1.8
cloudpickle==3.1.1
comm==0.2.2
contourpy==1.3.1
cramjam==2.9.1
cycler==0.12.1
debugpy==1.8.13
decorator==5.2.1
decord==0.6.0
defusedxml==0.7.1
diffusers==0.30.2
dm-tree==0.1.8
docker-pycreds==0.4.0
docstring_parser==0.16
einops==0.8.1
eval_type_backport==0.2.2
exceptiongroup==1.2.2
executing==2.2.0
Farama-Notifications==0.0.4
fastjsonschema==2.21.1
fastparquet==2024.11.0
filelock==3.18.0
flash-attn==2.7.1.post4
flatbuffers==25.2.10
fonttools==4.56.0
fqdn==1.5.1
frozenlist==1.5.0
fsspec==2025.3.0
fvcore==0.1.5.post20221221
gast==0.6.0
gitdb==4.0.12
GitPython==3.1.44
google-auth==2.38.0
google-auth-oauthlib==1.2.1
google-pasta==0.2.0
-e git+https://github.com/NVIDIA/Isaac-GR00T.git@0829414de5b28df8160be6b732972f12a2c04625#egg=gr00t
grpcio==1.71.0
gymnasium==1.0.0
h11==0.14.0
h5py==3.12.1
httpcore==1.0.7
httpx==0.28.1
huggingface-hub==0.29.3
hydra-core==1.3.2
idna==3.10
imageio==2.34.2
importlib_metadata==8.6.1
iniconfig==2.1.0
iopath==0.1.10
ipykernel==6.29.5
ipython==8.34.0
ipywidgets==8.1.5
isoduration==20.11.0
jedi==0.19.2
Jinja2==3.1.6
json5==0.10.0
jsonpointer==3.0.0
jsonschema==4.23.0
jsonschema-specifications==2024.10.1
jupyter==1.1.1
jupyter-console==6.6.3
jupyter-events==0.12.0
jupyter-lsp==2.2.5
jupyter_client==8.6.3
jupyter_core==5.7.2
jupyter_server==2.15.0
jupyter_server_terminals==0.5.3
jupyterlab==4.3.6
jupyterlab_pygments==0.3.0
jupyterlab_server==2.27.3
jupyterlab_widgets==3.0.13
keras==2.15.0
kiwisolver==1.4.8
kornia==0.7.4
kornia_rs==0.1.8
lazy_loader==0.4
libclang==18.1.1
llvmlite==0.44.0
Markdown==3.7
markdown-it-py==3.0.0
MarkupSafe==3.0.2
matplotlib==3.10.0
matplotlib-inline==0.1.7
mdurl==0.1.2
mistune==3.1.3
ml-dtypes==0.2.0
mpmath==1.3.0
msgpack==1.1.0
nbclient==0.10.2
nbconvert==7.16.6
nbformat==5.10.4
nest-asyncio==1.6.0
networkx==3.4.2
notebook==7.3.3
notebook_shim==0.2.4
numba==0.61.0
numpy==1.26.4
numpydantic==1.6.7
nvidia-cublas-cu12==12.4.5.8
nvidia-cuda-cupti-cu12==12.4.127
nvidia-cuda-nvrtc-cu12==12.4.127
nvidia-cuda-runtime-cu12==12.4.127
nvidia-cudnn-cu12==9.1.0.70
nvidia-cufft-cu12==11.2.1.3
nvidia-curand-cu12==10.3.5.147
nvidia-cusolver-cu12==11.6.1.9
nvidia-cusparse-cu12==12.3.1.170
nvidia-nccl-cu12==2.21.5
nvidia-nvjitlink-cu12==12.4.127
nvidia-nvtx-cu12==12.4.127
oauthlib==3.2.2
omegaconf==2.3.0
opencv-python==4.8.0.74
opencv-python-headless==4.11.0.86
opt_einsum==3.4.0
overrides==7.7.0
packaging==24.2
pandas==2.2.3
pandocfilters==1.5.1
parso==0.8.4
peft==0.14.0
pettingzoo==1.24.3
pexpect==4.9.0
pillow==11.1.0
pipablepytorch3d==0.7.6
platformdirs==4.3.7
pluggy==1.5.0
portalocker==3.1.1
prometheus_client==0.21.1
prompt_toolkit==3.0.50
protobuf==3.20.3
psutil==7.0.0
ptyprocess==0.7.0
pure_eval==0.2.3
pyarrow==14.0.1
pyasn1==0.6.1
pyasn1_modules==0.4.2
pyav==14.2.1
pycparser==2.22
pydantic==2.10.6
pydantic_core==2.27.2
Pygments==2.19.1
pyparsing==3.2.3
pytest==8.3.5
python-dateutil==2.9.0.post0
python-json-logger==3.3.0
pytz==2025.2
PyYAML==6.0.2
pyzmq==26.3.0
ray==2.40.0
referencing==0.36.2
regex==2024.11.6
requests==2.32.3
requests-oauthlib==2.0.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rich==13.9.4
rpds-py==0.24.0
rsa==4.9
safetensors==0.5.3
scikit-image==0.25.2
scipy==1.15.2
Send2Trash==1.8.3
sentry-sdk==2.24.1
setproctitle==1.3.5
shtab==1.7.1
six==1.17.0
smmap==5.0.2
sniffio==1.3.1
soupsieve==2.6
stack-data==0.6.3
sympy==1.13.1
tabulate==0.9.0
tensorboard==2.15.2
tensorboard-data-server==0.7.2
tensorflow==2.15.0
tensorflow-estimator==2.15.0
tensorflow-io-gcs-filesystem==0.37.1
termcolor==2.5.0
terminado==0.18.1
tianshou==0.5.1
tifffile==2025.3.13
timm==1.0.14
tinycss2==1.4.0
tokenizers==0.20.3
tomli==2.2.1
torch==2.5.1
torchvision==0.20.1
tornado==6.4.2
tqdm==4.67.1
traitlets==5.14.3
transformers==4.45.2
triton==3.1.0
typeguard==4.4.2
types-python-dateutil==2.9.0.20241206
typing_extensions==4.12.2
tyro==0.9.17
tzdata==2025.2
uri-template==1.3.0
urllib3==2.3.0
wandb==0.18.0
wcwidth==0.2.13
webcolors==24.11.1
webencodings==0.5.1
websocket-client==1.8.0
Werkzeug==3.1.3
widgetsnbextension==4.0.13
wrapt==1.14.1
yacs==0.1.8
zipp==3.21.0
zmq==0.0.0

My lerobot environment: 
Python 3.10.13
absl-py==2.1.0
aiohappyeyeballs==2.6.1
aiohttp==3.11.13
aiosignal==1.3.2
annotated-types==0.7.0
anthropic==0.49.0
antlr4-python3-runtime==4.9.3
anyio==4.9.0
apsw==3.49.1.0
apswutils==0.0.2
argon2-cffi==23.1.0
argon2-cffi-bindings==21.2.0
arrow==1.3.0
asciitree==0.3.3
asttokens==3.0.0
async-lru==2.0.5
async-timeout==5.0.1
attrs==25.3.0
av==14.2.0
babel==2.17.0
beautifulsoup4==4.13.3
bleach==6.2.0
blinker==1.9.0
certifi==2025.1.31
cffi==1.17.1
charset-normalizer==3.4.1
claudette==0.1.4
click==8.1.8
cloudpickle==3.1.1
cmake==3.31.6
comm==0.2.2
contourpy==1.3.1
cosette==0.0.4
cycler==0.12.1
datasets==3.4.0
debugpy==1.8.13
decorator==5.2.1
deepdiff==8.4.0
defusedxml==0.7.1
diffusers==0.32.2
dill==0.3.8
distro==1.9.0
dm-control==1.0.14
dm-env==1.6
dm-tree==0.1.9
docker-pycreds==0.4.0
draccus==0.10.0
einops==0.8.1
evdev==1.9.1
exceptiongroup==1.2.2
executing==2.2.0
Farama-Notifications==0.0.4
fastcore==1.7.29
fasteners==0.19
fastjsonschema==2.21.1
fastlite==0.1.2
feetech-servo-sdk==1.0.0
filelock==3.18.0
Flask==3.1.0
fonttools==4.56.0
fqdn==1.5.1
frozenlist==1.5.0
fsspec==2024.12.0
gdown==5.2.0
gitdb==4.0.12
GitPython==3.1.44
glfw==2.8.0
grpcio==1.71.0
gym-aloha==0.1.1
gym-pusht==0.1.5
gymnasium==0.29.1
h11==0.14.0
h5py==3.13.0
hf_transfer==0.1.9
html2text==2024.2.26
httpcore==1.0.7
httpx==0.28.1
huggingface-hub==0.29.3
hydra-core==1.3.2
idna==3.10
imageio==2.37.0
imageio-ffmpeg==0.6.0
importlib_metadata==8.6.1
iniconfig==2.0.0
inquirerpy==0.3.4
ipykernel==6.29.5
ipython==8.34.0
ipywidgets==8.1.5
isoduration==20.11.0
itsdangerous==2.2.0
jedi==0.19.2
Jinja2==3.1.6
jiter==0.9.0
json5==0.10.0
jsonlines==4.0.0
jsonpointer==3.0.0
jsonschema==4.23.0
jsonschema-specifications==2024.10.1
jupyter==1.1.1
jupyter-console==6.6.3
jupyter-events==0.12.0
jupyter-lsp==2.2.5
jupyter_client==8.6.3
jupyter_core==5.7.2
jupyter_server==2.15.0
jupyter_server_terminals==0.5.3
jupyterlab==4.3.6
jupyterlab_pygments==0.3.0
jupyterlab_server==2.27.3
jupyterlab_widgets==3.0.13
kiwisolver==1.4.8
lab==8.4
labmaze==1.0.6
lazy_loader==0.4
-e git+https://github.com/xuaner233/lerobot.git@b158576896fb57be0776e135b1c0f41a2e29d00a#egg=lerobot
llms-txt==0.0.4
llvmlite==0.44.0
lxml==5.3.1
Markdown==3.7
markdown-it-py==3.0.0
MarkupSafe==3.0.2
matplotlib==3.10.1
matplotlib-inline==0.1.7
mdurl==0.1.2
mergedeep==1.3.4
mistune==3.1.2
mpmath==1.3.0
msglm==0.0.6
mujoco==2.3.7
multidict==6.1.0
multiprocess==0.70.16
mypy-extensions==1.0.0
nbclient==0.10.2
nbconvert==7.16.6
nbformat==5.10.4
nest-asyncio==1.6.0
networkx==3.4.2
notebook==7.3.3
notebook_shim==0.2.4
numba==0.61.0
numcodecs==0.13.1
numpy @ file:///home/conda/feedstock_root/build_artifacts/numpy_1707225380409/work/dist/numpy-1.26.4-cp310-cp310-linux_x86_64.whl#sha256=51131fd8fc130cd168aecaf1bc0ea85f92e8ffebf211772ceb16ac2e7f10d7ca
nvidia-cublas-cu12==12.4.5.8
nvidia-cuda-cupti-cu12==12.4.127
nvidia-cuda-nvrtc-cu12==12.4.127
nvidia-cuda-runtime-cu12==12.4.127
nvidia-cudnn-cu12==9.1.0.70
nvidia-cufft-cu12==11.2.1.3
nvidia-curand-cu12==10.3.5.147
nvidia-cusolver-cu12==11.6.1.9
nvidia-cusparse-cu12==12.3.1.170
nvidia-cusparselt-cu12==0.6.2
nvidia-nccl-cu12==2.21.5
nvidia-nvjitlink-cu12==12.4.127
nvidia-nvtx-cu12==12.4.127
omegaconf==2.3.0
openai==1.66.3
opencv-python==4.10.0
opencv-python-headless==4.10.0
orderly-set==5.3.0
overrides==7.7.0
packaging==24.2
pandas==2.2.3
pandocfilters==1.5.1
parso==0.8.4
pexpect==4.9.0
pfzy==0.3.4
pillow==11.1.0
platformdirs==4.3.6
pluggy==1.5.0
prometheus_client==0.21.1
prompt_toolkit==3.0.50
propcache==0.3.0
protobuf==5.29.3
psutil==7.0.0
ptyprocess==0.7.0
pure_eval==0.2.3
pyarrow==19.0.1
pycparser==2.22
pydantic==2.10.6
pydantic_core==2.27.2
pygame==2.6.1
Pygments==2.19.1
pymunk==6.11.1
pynput==1.8.0
PyOpenGL==3.1.9
pyparsing==3.2.1
pyrealsense2==2.55.1.6486
pyserial==3.5
PySocks==1.7.1
pytest==8.3.5
python-dateutil==2.9.0.post0
python-json-logger==3.3.0
python-xlib==0.33
pytz==2025.1
PyYAML==6.0.2
pyyaml-include==1.4.1
pyzmq==26.3.0
referencing==0.36.2
regex==2024.11.6
requests==2.32.3
rerun-sdk==0.22.1
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rich==13.9.4
rpds-py==0.23.1
safetensors==0.5.3
scikit-image==0.25.2
scipy==1.15.2
Send2Trash==1.8.3
sentry-sdk==2.22.0
setproctitle==1.3.5
shapely==2.0.7
shell_sage==0.0.9
simplejson==3.20.1
six==1.17.0
smmap==5.0.2
sniffio==1.3.1
soupsieve==2.6
stack-data==0.6.3
sympy==1.13.1
tensorboard==2.19.0
tensorboard-data-server==0.7.2
termcolor==2.5.0
terminado==0.18.1
tifffile==2025.3.13
tinycss2==1.4.0
tokenizers==0.21.1
toml==0.10.2
tomli==2.2.1
toolslm==0.1.0
torch==2.6.0
TorchCodec==0.2.1
torchvision==0.21.0
tornado==6.4.2
tqdm==4.67.1
traitlets==5.14.3
transformers==4.49.0
triton==3.2.0
txt2tags==3.9
types-python-dateutil==2.9.0.20241206
typing-inspect==0.9.0
typing_extensions==4.12.2
tzdata==2025.1
uri-template==1.3.0
urllib3==2.3.0
wandb==0.19.8
wcwidth==0.2.13
webcolors==24.11.1
webencodings==0.5.1
websocket-client==1.8.0
Werkzeug==3.1.3
widgetsnbextension==4.0.13
wrapt==1.17.2
xxhash==3.5.0
yarl==1.18.3
zarr==2.18.3
zipp==3.21.0

**Comment by johnny555** [comment]
So it turns out that my model was just under fitted. 

After collecting a larger dataset (50 examples) and running on a large GPU (40GB VRAM) overnight I was able to get this working. I didn't use LoRA in the end.

Notably my grad norm was much smoother on the big GPU, I think training without LoRA is a good idea.

**Comment by youliangtan** [comment]
thanks for updating your results. Normally we would use at least 40 - 50 trajectories for post training, or even larger (> 100), depending on the tasks. Since so100 is an unseen embodiment, it normally requires more trajectories.

**Comment by VoicAndrei** [comment]
> So it turns out that my model was just under fitted.
> 
> After collecting a larger dataset (50 examples) and running on a large GPU (40GB VRAM) overnight I was able to get this working. I didn't use LoRA in the end.
> 
> Notably my grad norm was much smoother on the big GPU, I think training without LoRA is a good idea.

Did you use both cameras during the overnight training? Also, could you share your training parameters for the 40GB VRAM GPU? Thanks!
